lib/sqlalchemy/orm/session.py

```diff
@@ -1671,6 +1680,8 @@ def __init__(
         self.autoflush = autoflush
         self.expire_on_commit = expire_on_commit
         self.enable_baked_queries = enable_baked_queries
+        self.close_final = close_final
+        self._is_closed: bool = False
```

Comment:
**mike bayer** (zzzeek) wrote:

well this implies states, if I were doing that then I think we should make a real enum + state  ACTIVE / CLOSED_FINAL.   I dont think a "CLOSED" state is useful here.

we can make the whole thing a single state variable then:   self._closed_state = ClosedState.EMPTY_ONLY  / ClosedState.WILL_CLOSE_FINAL  / ClosedState.CLOSED_FINAL

@caselit names aside what do you think of a tri-state single variable like that?

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4824

Reply:
**mike bayer** (zzzeek) wrote:

we don't have an Enum for `Session` itself, we have two separate enums for SessionTransaction.  so this is a new enum.

"EMPTY_ONLY" is based on that other method I propose called `.empty()`, the reason for that name is that the basic thing `.close()` does is expunge all the objects as well as close any transaction.    so in that sense, "expunge_all_only"  maybe, but the closing of the transaction is still significant.

it seems like we are tying this verb "use" to the concept of `Session` until it's closed.  like we can run any number of operations on a `Session`, but as long as we dont call .close(), that's all one "use" ?

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4824

Reply:
**Federico Caselli** (CaselIT) wrote:

As mentioned in the other comment, I don't find empty clear at all. Maybe it's just me.

I've proposed alternative names that I think may be more familiar to users. https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4824/comment/10b8eeac_8db45620/

This enum is internal so I guess we can call things like we want, even `CLASSIC_MODE` works

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4824

Reply:
**Federico Caselli** (CaselIT) wrote:

Done

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4824